presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java

```diff
@@ -263,57 +268,87 @@ public void addSources(List<TaskSource> sources)
     {
         Map<PlanNodeId, TaskSource> updatedUnpartitionedSources = new HashMap<>();
 
-        // don't update maxAcknowledgedSplit until the end because task sources may not
-        // be in sorted order and if we updated early we could skip splits
-        long newMaxAcknowledgedSplit = maxAcknowledgedSplit;
+        // first remove any split that was already acknowledged
+        sources = sources.stream()
```

Comment:
I find it strange that `sources` is streamed through once, and then for-loop through once, and then streamed through once again. I would just put everything in one for-loop.